src/Elmish/Hooks/Type.purs

```diff
@@ -41,37 +41,33 @@ import Prelude as Prelude
 -- | because the first block has a `HookType` of `UseState String <> UseState
 -- | Int` and the second is `UseState Int <> UseState String`. The same hooks
 -- | need to be used in the same order for the hook types to match.
-foreign import data HookType :: Type
+foreign import data HookType' :: Type
+
+type HookType = HookType' -> HookType'
 
 -- | The `HookType` of `pure` — the identity of the `HookType` monoid.
-foreign import data Pure :: HookType
+foreign import data Pure :: HookType'
 
--- | A type which allows appending two `HookType`s.
-foreign import data AppendHookType :: HookType -> HookType -> HookType
+-- | A type which allows appending `HookType`s via composition.
+type AppendHookType (h1 :: HookType) (h2 :: HookType) t = h1 (h2 t)
```

Comment:
Waitasecond! I don't understand how this even works. A three-parameter type synonym can't be an infix operator, you should be getting "type synonym is partially applied".

Reply:
Maybe you haven't tried to compile any usages of  `<>` yet?

Reply:
But you’re allowed to partially apply types, right? Like, I can do this and it works fine:

```purs
type F a b = F a b
type H fa = H (F a)
```

Reply:
Ah, weird, I did run a build in `examples`, but am finally seeing errors when I run `npm start` in `examples`.

Reply:
Well, that’s a shame, but we can live with:

```purs
type UseLocalStorage t = UseState String <> UseEffect Unit <> t

useLocalStorage :: String -> String -> Hook UseLocalStorage (String /\ Dispatch String)
```

Reply:
That particular example doesn't even make sense: `type F a b = F a b` is a diverging recursive definition. Type `F` is defined as itself.

If you replace `type` with `data`, that still doesn't work, because `F a` is definitely not a `Type`, so it can't be wrapped in constructor `H`.

But more generally - yes, you can partially apply types, but not _type synonyms_. Try this:

```
data F a b = F a b
type G = F Int -- works, because F is `data`

type F' a b = F a b
type G' = F' Int -- error: partially applied type synonym
```

This has to do with some rather subtle problems that occur if you allowed that. Basically type safety breaks.

Reply:
Yes, we can live with that. I think that's good enough, all ugliness is only visible to hook authors.